test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +519,682 @@ func TestRSSelectorImmutability(t *testing.T) {
+		t.Fatalf("failed to update pod %s: %v", pod.Name, err)
+	}
+	pod, err = c.Core().Pods(ns.Name).Get(pod.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get pod %s: %v", pod.Name, err)
+	}
+	// Verify the RS does not adopt the pod automatically after reverting back the selector
+	if len(pod.OwnerReferences) != 0 {
+		t.Fatalf("pod %s was not expected to be adopted by RS %s", pod.Name, rs.Name)
+	}
+
+	// We have to set OwnerReferences of the pod manually for it to be adopted by the RS
+	var trueVar = true
+	controllerReference := metav1.OwnerReference{UID: rs.UID, APIVersion: "v1beta1", Kind: "ReplicaSet", Name: rs.Name, Controller: &trueVar}
+	pod.OwnerReferences = []metav1.OwnerReference{controllerReference}
+
+	// Have to GET the pod again as the pod has been modified
+	pod, err = c.Core().Pods(ns.Name).Get(pod.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get pod %s: %v", pod.Name, err)
+	}
+	pod, err = c.Core().Pods(ns.Name).Update(pod)
+	if err != nil {
+		t.Fatalf("failed to update pod %s: %v", pod.Name, err)
+	}
+	pod, err = c.Core().Pods(ns.Name).Get(pod.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get pod %s: %v", pod.Name, err)
+	}
+	// Perhaps a better way is to compare the OwnerReferences using reflect.DeepEqual (this comment will be removed afterwards)
+	if len(pod.OwnerReferences) != 1 {
+		t.Fatalf("failed to add controllerReference for pod %s after explicitly setting its OwnerReferences to RS %s", pod.Name, rs.Name)
+	}
+}
+
+func TestPatchPod(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 1)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	pod := newMatchingPod("pod", ns.Name)
+	pod.Labels = labelMap
+	createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod}, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	var trueVar = true
+	var falseVar = false
+	expectedControllerReference := metav1.OwnerReference{UID: rs.UID, APIVersion: "extensions/v1beta1", Kind: "ReplicaSet", Name: rs.Name, Controller: &trueVar}
+	expectedOwnerReferences := []metav1.OwnerReference{expectedControllerReference}
+
+	// When the only OwnerReference of the pod points to another type of API object such as statefulset
+	// with Controller=false, the RS should add a second OwnerReference pointing to itself
+	// with Controller=true
+	controllerReference := metav1.OwnerReference{UID: uuid.NewUUID(), APIVersion: "apps/v1beta1", Kind: "StatefulSet", Name: rs.Name, Controller: &falseVar}
+	pod.OwnerReferences = []metav1.OwnerReference{controllerReference}
+	pod, err := c.Core().Pods(ns.Name).Get(pod.Name, metav1.GetOptions{})
```

Comment:
Actually, every time you do Get, modify, Update for any object, you should be using a helper that does retries on conflict, or else the test will be flaky since you may race with the controller. There is a helper called `RetryOnConflict` in client-go/util/retry that would be useful here.

Reply:
Yes, the mutation should be between Get and Update. In this case, GET is unnecessary.
I will search for `RetryOnConflict` helper function to prevent racing with the controller when do Get, modify, Update.